src/server/session.ts

```diff
@@ -2795,6 +2796,26 @@ export class Session<TMessage = string> implements EventSender {
         return project.getLanguageService().getMoveToRefactoringFileSuggestions(file, this.extractPositionOrRange(args, scriptInfo), this.getPreferences(file));
     }
 
+    private getPasteEdits(args: protocol.GetPasteEditsRequestArgs): protocol.PasteEditsAction | undefined {
+        const { file, project } = this.getFileAndProject(args);
+        const copyFile = args.copies[0].range ? args.copies[0].range.file : undefined;
+        const result = project.getLanguageService().getPasteEdits(
+            file,
+            args.copies.map(copy => ({
+                text: copy.text,
+                copyRange: copy.range && copyFile
+                    ? { file: copy.range.file, range: this.getRange({ file: copyFile, startLine: copy.range.start.line, startOffset: copy.range.start.offset, endLine: copy.range.end.line, endOffset: copy.range.end.offset }, project.getScriptInfoForNormalizedPath(toNormalizedPath(copyFile))!) } : undefined,

```

Comment:
You’re assuming, but not checking, that every copy has the same file. I think this is necessarily true from VS Code, but the protocol suggests that they could be different. We should revise the protocol to fix that, handle differing files, or debug assert that all copies have the same file.